File: src/tools/projects.ts
import { McpServer } from "@mcp/server";
import { z } from "zod";
import type { LeantimeClient } from "../leantime-client.ts";
import { markdownToHtml } from "../markdown.ts";
import { errorResult, isLeantimeError, MARKDOWN_HINT } from "./shared.ts";

export function registerProjectTools(server: McpServer, client: LeantimeClient) {
  server.tool(
    "leantime_list_projects",
    "List all projects assigned to the current user",
    {},
    async () => {
      try {
        const result = await client.call<Record<string, unknown>[]>(
          "Projects.getAll",
        );
        return {
          content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_get_project",
    "Get details of a specific project",
    {
      projectId: z.string().describe("The project ID"),
    },
    async ({ projectId }) => {
      try {
        const result = await client.call<Record<string, unknown>>(
          "projects.getProject",
          { id: projectId },
        );
        return {
          content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_get_project_progress",
    "Get progress metrics for a specific project",
    {
      projectId: z.string().describe("The project ID"),
    },
    async ({ projectId }) => {
      try {
        const result = await client.call<Record<string, unknown>>(
          "projects.getProjectProgress",
          { projectId },
        );
        return {
          content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_create_project",
    `Create a new project. The details field is ${MARKDOWN_HINT}. ` +
      `Get a valid clientId with leantime_list_clients first.`,
    {
      name: z.string().describe("Project name"),
      clientId: z.string().describe("Client ID (see leantime_list_clients)"),
      details: z.string().optional().describe(`Project details in ${MARKDOWN_HINT}`),
      hourBudget: z.number().optional().describe("Hour budget"),
      dollarBudget: z.number().optional().describe("Dollar budget"),
    },
    async ({ name, clientId, details, hourBudget, dollarBudget }) => {
      try {
        const values: Record<string, unknown> = { name, clientId };
        if (details) values.details = markdownToHtml(details);
        if (hourBudget !== undefined) values.hourBudget = hourBudget;
        if (dollarBudget !== undefined) values.dollarBudget = dollarBudget;

        const result = await client.call<unknown>("projects.addProject", { values });
        if (result === false || result === null || isLeantimeError(result)) {
          return errorResult("Project creation failed.");
        }
        // addProject returns the new id wrapped in an array — normalize.
        const normalized = Array.isArray(result) ? { id: result[0] } : { id: result };
        return {
          content: [{ type: "text", text: JSON.stringify(normalized, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_update_project",
    `Update a project. Only the provided fields are changed (patch API — other ` +
      `fields are never wiped). The details field is ${MARKDOWN_HINT}.`,
    {
      projectId: z.string().describe("The project ID"),
      name: z.string().optional().describe("New project name"),
      details: z.string().optional().describe(`New details in ${MARKDOWN_HINT}`),
      hourBudget: z.number().optional().describe("New hour budget"),
      dollarBudget: z.number().optional().describe("New dollar budget"),
    },
    async ({ projectId, name, details, hourBudget, dollarBudget }) => {
      try {
        const params: Record<string, unknown> = {};
        if (name !== undefined) params.name = name;
        if (details !== undefined) params.details = markdownToHtml(details);
        if (hourBudget !== undefined) params.hourBudget = hourBudget;
        if (dollarBudget !== undefined) params.dollarBudget = dollarBudget;

        if (Object.keys(params).length === 0) {
          return errorResult("Nothing to update: provide at least one field to change.");
        }

        const result = await client.call<boolean>("projects.patch", {
          id: projectId,
          params,
        });
        return {
          content: [{
            type: "text",
            text: JSON.stringify({ ok: result === true, id: projectId }, null, 2),
          }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_find_projects",
    "Search projects by name (fuzzy)",
    {
      term: z.string().describe("Search term"),
    },
    async ({ term }) => {
      try {
        const result = await client.call<Record<string, unknown>[]>(
          "projects.findProject",
          { term },
        );
        // findProject mangles ids into "id-modified" — normalize back to plain ids.
        const normalized = (result ?? []).map((p) => ({
          ...p,
          id: String(p.id).split("-")[0],
        }));
        return {
          content: [{ type: "text", text: JSON.stringify(normalized, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_list_project_users",
    "List the users assigned to a project (id, name) — the valid editorId " +
      "candidates for tickets and milestones of that project",
    {
      projectId: z.string().describe("The project ID"),
    },
    async ({ projectId }) => {
      try {
        const users = await client.call<Record<string, unknown>[]>(
          "projects.getUsersAssignedToProject",
          { projectId },
        );
        const simplified = (users ?? []).map((u) => ({
          id: String(u.id),
          name: [u.firstname, u.lastname].filter(Boolean).join(" ").trim(),
          role: u.projectRole ?? undefined,
        }));
        return {
          content: [{ type: "text", text: JSON.stringify(simplified, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );

  server.tool(
    "leantime_list_clients",
    "List all clients (id, name) — clientId is required to create projects",
    {},
    async () => {
      try {
        const clients = await client.call<Record<string, unknown>[]>("Clients.getAll", {});
        const simplified = (clients ?? []).map((c) => ({
          id: String(c.id),
          name: c.name,
        }));
        return {
          content: [{ type: "text", text: JSON.stringify(simplified, null, 2) }],
        };
      } catch (e) {
        return errorResult(e instanceof Error ? e.message : String(e));
      }
    },
  );
}

File: src/tools/projects.rs
use std::future::Future;
use std::pin::Pin;

use serde_json::{json, Value};

use crate::markdown::markdown_to_html;

use super::shared::*;
use super::{error_result, ok_result, rpc, ClientRef, Tool, ToolAnnotations};

fn h_list_projects(_a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        match c.call("Projects.getAll", json!({})).await {
            Ok(r) => ok_result(&r),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_get_project(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let id = a.get("projectId").and_then(|v| v.as_str()).unwrap_or("");
        match c.call("projects.getProject", json!({"id": id})).await {
            Ok(r) => ok_result(&r),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_create_project(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let mut v = json!({"name": a.get("name"), "clientId": a.get("clientId")});
        if let Some(d) = a.get("details").and_then(|x| x.as_str()) {
            v["details"] = json!(markdown_to_html(d));
        }
        if let Some(x) = a.get("hourBudget") {
            v["hourBudget"] = x.clone();
        }
        if let Some(x) = a.get("dollarBudget") {
            v["dollarBudget"] = x.clone();
        }
        match c.call("projects.addProject", json!({"values": v})).await {
            Ok(r) => {
                if is_leantime_error(&r) {
                    return error_result(&leantime_error_msg(&r));
                }
                if r == json!(false) || r.is_null() {
                    return error_result("Project creation failed.");
                }
                let id = r.as_array().and_then(|x| x.first()).cloned().unwrap_or(r);
                ok_result(&json!({ "id": id }))
            }
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_update_project(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let id = a.get("projectId").and_then(|v| v.as_str()).unwrap_or("");
        let mut params = json!({});
        if let Some(x) = a.get("name") {
            params["name"] = x.clone();
        }
        if let Some(d) = a.get("details").and_then(|x| x.as_str()) {
            params["details"] = json!(markdown_to_html(d));
        }
        if let Some(x) = a.get("hourBudget") {
            params["hourBudget"] = x.clone();
        }
        if let Some(x) = a.get("dollarBudget") {
            params["dollarBudget"] = x.clone();
        }
        if params.as_object().map(|o| o.is_empty()).unwrap_or(true) {
            return error_result("Nothing to update: provide at least one field to change.");
        }
        match c
            .call("projects.patch", json!({"id": id, "params": params}))
            .await
        {
            Ok(r) => ok_result(&json!({ "ok": r == json!(true), "id": id })),
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_find_projects(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let term = a.get("term").and_then(|v| v.as_str()).unwrap_or("");
        match c.call("projects.findProject", json!({"term": term})).await {
            Ok(r) => {
                // findProject mangles ids into "id-modified" — normalize back to plain ids.
                let mut items = r.as_array().cloned().unwrap_or_default();
                for p in items.iter_mut() {
                    if let Some(idv) = p.get("id") {
                        let raw = match idv {
                            Value::String(s) => s.clone(),
                            other => other.to_string(),
                        };
                        if let Some(obj) = p.as_object_mut() {
                            obj.insert("id".into(), json!(raw.split('-').next().unwrap_or(&raw)));
                        }
                    }
                }
                ok_result(&json!(items))
            }
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_list_project_users(a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        let pid = a.get("projectId").and_then(|v| v.as_str()).unwrap_or("");
        match c
            .call(
                "projects.getUsersAssignedToProject",
                json!({"projectId": pid}),
            )
            .await
        {
            Ok(r) => {
                let simplified: Vec<Value> = r.as_array().map(|arr| arr.iter().map(|u| {
                    let mut o = json!({
                        "id": u.get("id").map(|v| match v { Value::String(s) => json!(s), other => json!(other.to_string()) }).unwrap_or(json!("")),
                        "name": format!("{} {}",
                            u.get("firstname").and_then(|v| v.as_str()).unwrap_or(""),
                            u.get("lastname").and_then(|v| v.as_str()).unwrap_or("")).trim(),
                    });
                    if let Some(role) = u.get("projectRole") { if !role.is_null() { o["role"] = role.clone(); } }
                    o
                }).collect()).unwrap_or_default();
                ok_result(&json!(simplified))
            }
            Err(e) => error_result(&e.to_string()),
        }
    })
}

fn h_list_clients(_a: Value, cl: ClientRef) -> Pin<Box<dyn Future<Output = Value> + Send>> {
    Box::pin(async move {
        let mut c = cl.lock().await;
        match c.call("Clients.getAll", json!({})).await {
            Ok(r) => {
                let simplified: Vec<Value> = r.as_array().map(|arr| arr.iter().map(|cl| json!({
                    "id": cl.get("id").map(|v| match v { Value::String(s) => json!(s), other => json!(other.to_string()) }).unwrap_or(json!("")),
                    "name": cl.get("name").cloned().unwrap_or(Value::Null),
                })).collect()).unwrap_or_default();
                ok_result(&json!(simplified))
            }
            Err(e) => error_result(&e.to_string()),
        }
    })
}

pub(super) fn tools() -> Vec<Tool> {
    let md = MARKDOWN_HINT;
    vec![
        tool("leantime_list_projects", "List all projects assigned to the current user",
            vec![], vec![], Box::new(h_list_projects)),
        tool("leantime_get_project", "Get details of a specific project",
            vec![rs("projectId", "The project ID")], vec!["projectId"], Box::new(h_get_project)),
        tool("leantime_get_project_progress", "Get progress metrics for a specific project",
            vec![rs("projectId", "The project ID")], vec!["projectId"],
            rpc("projects.getProjectProgress", |a: &Value| json!({"projectId": a.get("projectId")}))),
        tool_with_annotations("leantime_create_project", format!("Create a new project. The details field is {}. Get a valid clientId with leantime_list_clients first.", md),
            vec![rs("name", "Project name"), rs("clientId", "Client ID (see leantime_list_clients)"), os("details", format!("Project details in {}", md)), on("hourBudget", "Hour budget"), on("dollarBudget", "Dollar budget")],
            vec!["name", "clientId"], Box::new(h_create_project), ToolAnnotations::write()),
        tool_with_annotations("leantime_update_project", "Update a project. Only the provided fields are changed (patch API — other fields are never wiped).",
            vec![rs("projectId", "The project ID"), os("name", "New project name"), os("details", format!("New details in {}", md)), on("hourBudget", "New hour budget"), on("dollarBudget", "New dollar budget")],
            vec!["projectId"], Box::new(h_update_project), ToolAnnotations::write()),
        tool("leantime_find_projects", "Search projects by name (fuzzy)",
            vec![rs("term", "Search term")], vec!["term"], Box::new(h_find_projects)),
        tool("leantime_list_project_users", "List the users assigned to a project (id, name) — the valid editorId candidates for tickets and milestones of that project",
            vec![rs("projectId", "The project ID")], vec!["projectId"], Box::new(h_list_project_users)),
        tool("leantime_list_clients", "List all clients (id, name) — clientId is required to create projects",
            vec![], vec![], Box::new(h_list_clients)),
    ]
}
